File: src/main/java/com/ultreon/ultranlang/Interpreter.java
package com.ultreon.ultranlang;

import com.ultreon.ultranlang.annotations.Visit;
import com.ultreon.ultranlang.ast.*;
import com.ultreon.ultranlang.symbol.FuncSymbol;
import com.ultreon.ultranlang.symbol.VarSymbol;
import com.ultreon.ultranlang.token.TokenType;
import com.ultreon.ultranlang.utils.CollectionUtils;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.Nullable;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.math.RoundingMode;
import java.util.List;
import java.util.Map;

@SuppressWarnings("SpellCheckingInspection")
public class Interpreter extends NodeVisitor {
    @Nullable
    private final Program tree;
    private final CallStack callStack = new CallStack();

    public Interpreter(@Nullable Program tree) {
        this.tree = tree;
    }

    public void log(Object msg) {
        if (Spi.SHOULD_LOG_STACK) {
            System.out.println(msg);
        }
    }

    @Visit(Program.class)
    public void visitProgram(Program node) {
        String programName = node.getName();
        log("ENTER: PROGRAM " + programName);

        ActivationRecord ar = new ActivationRecord(programName, ARType.PROGRAM, 1);
        callStack.push(ar);

        log(callStack.toString());

        for (AST statement : node.getStatements()) {
            visit(statement);
        }
    }

    @Visit(Block.class)
    public void visitBlock(Block node) {
        for (AST declaration : node.getDeclarations()) {
            visit(declaration);
        }
        visit(node.getCompoundStatement());
    }

    @Visit(VarDecl.class)
    public void visitVarDecl(VarDecl node) {
        // do nothing
    }

    @Visit(Type.class)
    public void visitType(Type node) {
        // do nothing
    }

    @Nullable
    @Visit(BinOp.class)
    public Object visitBinOp(BinOp node) {
        if (node.getOp().getType().equals(TokenType.PLUS)) {
            return plus(visit(node.getLeft()), visit(node.getRight()));
        } else if (node.getOp().getType().equals(TokenType.MINUS)) {
            return minus(visit(node.getLeft()), visit(node.getRight()));
        } else if (node.getOp().getType().equals(TokenType.MUL)) {
            return mul(visit(node.getLeft()), visit(node.getRight()));
        } else if (node.getOp().getType().equals(TokenType.INTEGER_DIV)) {
            return div(visit(node.getLeft()), visit(node.getRight()));
        } else if (node.getOp().getType().equals(TokenType.FLOAT_DIV)) {
            return div(visit(node.getLeft()), visit(node.getRight()));
//        } else if (node.getOp().getType().equals(TokenType.EQ)) {
//            return eq(visit(node.getLeft()), visit(node.getRight()));
//        } else if (node.getOp().getType().equals(TokenType.NEQ)) {
//            return neq(visit(node.getLeft()), visit(node.getRight()));
//        } else if (node.getOp().getType().equals(TokenType.LT)) {
//            return lt(visit(node.getLeft()), visit(node.getRight()));
//        } else if (node.getOp().getType().equals(TokenType.LE)) {
//            return le(visit(node.getLeft()), visit(node.getRight()));
//        } else if (node.getOp().getType().equals(TokenType.GT)) {
//            return gt(visit(node.getLeft()), visit(node.getRight()));
//        } else if (node.getOp().getType().equals(TokenType.GE)) {
//            return ge(visit(node.getLeft()), visit(node.getRight()));
        } else {
            throw new RuntimeException("Unknown operator: " + node.getOp().getType());
        }
    }

    @Visit(Num.class)
    public Object visitNum(Num node) {
        return node.getValue();
    }

    @Visit(Str.class)
    public Object visitStr(Str node) {
        return node.getValue();
    }

    @Visit(UnaryOp.class)
    public Object visitUnaryOp(UnaryOp node) {
        if (node.getOp().getType().equals(TokenType.PLUS)) {
            return unaryPlus(visit(node.getExpr()));
        } else if (node.getOp().getType().equals(TokenType.MINUS)) {
            return unaryMinus(visit(node.getExpr()));
//        } else if (node.getOp().getType().equals(TokenType.NOT)) {
//            return not(visit(node.getExpr()));
        } else {
            throw new RuntimeException("Unknown operator: " + node.getOp().getType());
        }
    }

    @Nullable
    @Contract("_->null")
    @Visit(Compound.class)
    public Object visitCompound(Compound node) {
        for (AST child : node.getChildren()) {
            visit(child);
        }
        return null;
    }

    @Visit(Assign.class)
    public Object visitAssign(Assign node) {
        String varName = (String) node.getLeft().getValue();
        Object value = visit(node.getRight());

        ActivationRecord ar = callStack.peek();
        ar.set(varName, value);

        // TODO: return set value
        return null;
    }

    @Visit(Var.class)
    public Object visitVar(Var node) {
        String varName = (String) node.getValue();

        ActivationRecord ar = callStack.peek();
        return ar.get(varName);
    }

    @Visit(NoOp.class)
    public void visitNoOp(NoOp node) {
        // do nothing
    }

    @Visit(FuncDeclaration.class)
    public void visitFuncDeclaration(FuncDeclaration node) {

    }

    @Visit(FuncCall.class)
    @SuppressWarnings("ConstantConditions")
    public Object visitFuncCall(FuncCall node) {
        String funcName = node.getFuncName();
        FuncSymbol funcSymbol = node.getFuncSymbol();

        ActivationRecord ar = new ActivationRecord(funcName, ARType.FUNCTION, funcSymbol.getScopeLevel() + 1);

        List<VarSymbol> formalParams = funcSymbol.getFormalParams();
        List<AST> actualParams = node.getActualParams();

        for (Map.Entry<VarSymbol, AST> entry : CollectionUtils.zip(formalParams, actualParams).entrySet()) {
            ar.set(entry.getKey().getName(), visit(entry.getValue()));
        }

        this.callStack.push(ar);

        this.log("ENTER: FUNCTION " + funcName);
        this.log(callStack.toString());

        // evaluate function body
        Object value;

        if (funcSymbol.isNative()) {
            Object returned = funcSymbol.callNative(ar);
            if (returned instanceof Void) {
                returned = null;
            }
            value = returned;
        } else {
            for (AST statement : funcSymbol.getStatements()) {
                visit(statement);
            }
            value = null;
        }

        this.log("LEAVE: FUNCTION " + funcName);
        this.log(callStack.toString());

        this.callStack.pop();

        return value;
    }

    public boolean interpret() {
        if (this.tree == null) {
            return false;
        }
        this.visit(tree);
        return true;
    }

    private Object plus(Object a, Object b) {
        if (a instanceof BigInteger && b instanceof BigInteger) {
            return ((BigInteger) a).add((BigInteger) b);
        } else if (a instanceof BigDecimal && b instanceof BigDecimal) {
            return ((BigDecimal) a).add((BigDecimal) b);
        } else if (a instanceof BigInteger && b instanceof BigDecimal) {
            return new BigDecimal((BigInteger) a).add((BigDecimal) b);
        } else if (a instanceof BigDecimal && b instanceof BigInteger) {
            return ((BigDecimal) a).add(new BigDecimal(((BigInteger) b)));
        } else if (a instanceof String && b instanceof String) {
            return a + (String) b;
        } else if (a != null && b instanceof String) {
            return a.toString() + b;
        } else if (a instanceof String && b != null) {
            return a + b.toString();
        } else {
            throw new RuntimeException("Cannot add " + (a == null ? "null" : a.getClass().getSimpleName()) + " and " + (b == null ? "null" : b.getClass().getSimpleName()));
        }
    }

    private Object minus(Object a, Object b) {
        if (a instanceof BigInteger && b instanceof BigInteger) {
            return ((BigInteger) a).subtract((BigInteger) b);
        } else if (a instanceof BigDecimal && b instanceof BigDecimal) {
            return ((BigDecimal) a).subtract((BigDecimal) b);
        } else if (a instanceof BigInteger && b instanceof BigDecimal) {
            return new BigDecimal((BigInteger) a).subtract((BigDecimal) b);
        } else if (a instanceof BigDecimal && b instanceof BigInteger) {
            return ((BigDecimal) a).subtract(new BigDecimal(((BigInteger) b)));
        } else {
            throw new RuntimeException("Cannot subtract " + (a == null ? "null" : a.getClass().getSimpleName()) + " and " + (b == null ? "null" : b.getClass().getSimpleName()));
        }
    }

    private Object mul(Object a, Object b) {
        if (a instanceof BigInteger && b instanceof BigInteger) {
            return ((BigInteger) a).multiply((BigInteger) b);
        } else if (a instanceof BigDecimal && b instanceof BigDecimal) {
            return ((BigDecimal) a).multiply((BigDecimal) b);
        } else if (a instanceof BigInteger && b instanceof BigDecimal) {
            return new BigDecimal((BigInteger) a).multiply((BigDecimal) b);
        } else if (a instanceof BigDecimal && b instanceof BigInteger) {
            return ((BigDecimal) a).multiply(new BigDecimal(((BigInteger) b)));
        } else {
            throw new RuntimeException("Cannot multiply " + (a == null ? "null" : a.getClass().getSimpleName()) + " and " + (b == null ? "null" : b.getClass().getSimpleName()));
        }
    }

    private Object div(Object a, Object b) {
        if (a instanceof BigInteger && b instanceof BigInteger) {
            return ((BigInteger) a).divide((BigInteger) b);
        } else if (a instanceof BigDecimal && b instanceof BigDecimal) {
            return ((BigDecimal) a).divide((BigDecimal) b, RoundingMode.HALF_UP);
        } else if (a instanceof BigInteger && b instanceof BigDecimal) {
            return new BigDecimal((BigInteger) a).divide((BigDecimal) b, RoundingMode.HALF_UP);
        } else if (a instanceof BigDecimal && b instanceof BigInteger) {
            return ((BigDecimal) a).divide(new BigDecimal(((BigInteger) b)), RoundingMode.HALF_UP);
        } else {
            throw new RuntimeException("Cannot divide " + (a == null ? "null" : a.getClass().getSimpleName()) + " and " + (b == null ? "null" : b.getClass().getSimpleName()));
        }
    }

    private Object unaryPlus(Object a) {
        if (a instanceof BigInteger) {
            return ((BigInteger) a).abs();
        } else if (a instanceof BigDecimal) {
            return ((BigDecimal) a).abs();
        } else {
            throw new RuntimeException("Cannot unary plus " + (a == null ? "null" : a.getClass().getSimpleName()));
        }
    }

    private Object unaryMinus(Object a) {
        if (a instanceof BigInteger) {
            return ((BigInteger) a).negate();
        } else if (a instanceof BigDecimal) {
            return ((BigDecimal) a).negate();
        } else {
            throw new RuntimeException("Cannot unary minus " + (a == null ? "null" : a.getClass().getSimpleName()));
        }
    }
}

File: src/main/kotlin/com/ultreon/ultranlang/Interpreter.kt
package com.ultreon.ultranlang

import com.ultreon.ultranlang.annotations.Visit
import com.ultreon.ultranlang.ast.*
import com.ultreon.ultranlang.token.TokenType
import java.math.BigDecimal
import java.math.BigInteger

class Interpreter(val tree: Program?) : NodeVisitor() {
    val callStack: CallStack = CallStack()
    
    fun log(msg: Any?) {
        if (shouldLogStack) {
            println(msg)
        }
    }
    
    @Visit(Program::class)
    fun visitProgram(node: Program) {
        val programName = node.name
        log("ENTER: PROGRAM $programName")

        val ar = ActivationRecord(programName, ARType.PROGRAM, 1)
        callStack.push(ar)

        log(callStack.toString())

        for (statement in node.statements) {
            visit(statement)
        }

        log("LEAVE: PROGRAM $programName")
        log(callStack.toString())

        callStack.pop()
    }
    
    @Visit(Block::class)
    fun visitBlock(node: Block) {
        for (declaration in node.declarations) {
            visit(declaration)
        }
        
        visit(node.compoundStatement)
    }
    
    @Visit(VarDecl::class)
    fun visitVarDecl(node: VarDecl) {
        // do nothing
    }
    
    @Visit(Type::class)
    fun visitType(node: Type) {
        // do nothing
    }
    
    @Visit(BinOp::class)
    fun visitBinOp(node: BinOp): Any? {
        return when (node.op.type) {
            TokenType.PLUS -> visit(node.left) + visit(node.right)
            TokenType.MINUS -> visit(node.left) - visit(node.right)
            TokenType.MUL -> visit(node.left) * visit(node.right)
            TokenType.INTEGER_DIV -> visit(node.left) / visit(node.right)
            else -> throw IllegalArgumentException("Unknown operator ${node.op.type}")
        }
    }

    @Visit(Num::class)
    fun visitNum(node: Num): Any? {
        return node.value
    }

    @Visit(com.ultreon.ultranlang.ast.String::class)
    fun visitString(node: com.ultreon.ultranlang.ast.String): Any? {
        return node.value
    }

    @Visit(UnaryOp::class)
    fun visitUnaryOp(node: UnaryOp): Any? {
        return when (node.op.type) {
            TokenType.PLUS -> +visit(node.expr)
            TokenType.MINUS -> -visit(node.expr)
            else -> throw IllegalArgumentException("Unknown operator ${node.op.type}")
        }
    }

    @Visit(Compound::class)
    fun visitCompound(node: Compound): Any? {
        for (child in node.children) {
            visit(child)
        }

        return null
    }

    @Visit(Assign::class)
    fun visitAssign(node: Assign): Any? {
        val varName = node.left.value as String
        val varValue = visit(node.right)

        val ar = callStack.peek()
        ar[varName] = varValue

        return null
    }

    @Visit(Var::class)
    fun visitVar(node: Var): Any? {
        val varName = node.value as String

        val ar = callStack.peek()
        return ar[varName]
    }

    @Visit(NoOp::class)
    fun visitNoOp(node: NoOp) {
        // do nothing
    }

    @Visit(FuncDeclaration::class)
    fun visitProcedureDecl(node: FuncDeclaration) {
        // do nothing
    }

    @Visit(FuncCall::class)
    fun visitProcedureCall(node: FuncCall): Any? {
        val procName = node.procName
        val procSymbol = node.procSymbol

        val ar = ActivationRecord(procName, ARType.PROCEDURE, procSymbol!!.scopeLevel + 1)

        val formalParams = procSymbol.formalParams
        val actualParams = node.actualParams

        for ((paramSymbol, argumentNode) in formalParams.zip(actualParams)) {
            ar[paramSymbol.name] = this.visit(argumentNode)
        }

        this.callStack.push(ar)

        this.log("ENTER: PROCEDURE $procName")
        this.log(this.callStack.toString())

        // evaluate procedure body
        val value: Any?

        if (procSymbol.isNative) {
            var returned = procSymbol.callNative(ar)
            if (returned == Unit) {
                returned = null
            }
            value = returned
        } else {
            for (statement in procSymbol.statements) {
                this.visit(statement)
            }
            value = null
        }

        this.log("LEAVE: PROCEDURE $procName")
        this.log(this.callStack.toString())

        this.callStack.pop()
        return value
    }

    fun interpret(): Any {
        val tree = this.tree ?: return ""

        val visit = this.visit(tree)
        return visit as Unit
    }
}

private operator fun Any?.unaryPlus(): Any? {
    return when (this) {
        is Byte -> +this
        is Short -> +this
        is Int -> +this
        is Long -> +this
        is Float -> +this
        is Double -> +this
        is BigInteger -> this.abs()
        is BigDecimal -> this.abs()
        else -> throw IllegalArgumentException("Unknown unary operator +")
    }
}

private operator fun Any?.unaryMinus(): Any? {
    return when (this) {
        is Byte -> +this
        is Short -> +this
        is Int -> +this
        is Long -> +this
        is Float -> +this
        is Double -> +this
        is BigInteger -> this.negate()
        is BigDecimal -> this.negate()
        else -> throw IllegalArgumentException("Unknown unary operator +")
    }
}

private operator fun Any?.plus(visit: Any?): Any? {
    return if (this is Byte && visit is Byte) {
        this + visit
    } else if (this is Short && visit is Short) {
        this + visit
    } else if (this is Int && visit is Int) {
        this + visit
    } else if (this is Long && visit is Long) {
        this + visit
    } else if (this is BigInteger && visit is BigInteger) {
        this.add(visit)
    } else if (this is BigDecimal && visit is BigDecimal) {
        this.add(visit)
    } else if (this is String && visit is String) {
        this + visit
    } else if (this is String && visit is Any) {
        this + visit
    } else if (this is Any && visit is String) {
        this.toString() + visit
    } else {
        throw IllegalArgumentException("Cannot add $this and $visit")
    }
}

private operator fun Any?.minus(visit: Any?): Any? {
    return if (this is Byte && visit is Byte) {
        this - visit
    } else if (this is Short && visit is Short) {
        this - visit
    } else if (this is Int && visit is Int) {
        this - visit
    } else if (this is Long && visit is Long) {
        this - visit
    } else if (this is BigInteger && visit is BigInteger) {
        this.subtract(visit)
    } else if (this is BigDecimal && visit is BigDecimal) {
        this.subtract(visit)
    } else if (this is String && visit is String) {
        this - visit
    } else if (this is String && visit is Any) {
        this - visit
    } else if (this is Any && visit is String) {
        this.toString() - visit
    } else {
        throw IllegalArgumentException("Cannot subtract $this and $visit")
    }
}

private operator fun Any?.times(visit: Any?): Any? {
    return if (this is Byte && visit is Byte) {
        this * visit
    } else if (this is Short && visit is Short) {
        this * visit
    } else if (this is Int && visit is Int) {
        this * visit
    } else if (this is Long && visit is Long) {
        this * visit
    } else if (this is BigInteger && visit is BigInteger) {
        this.multiply(visit)
    } else if (this is BigDecimal && visit is BigDecimal) {
        this.multiply(visit)
    } else {
        throw IllegalArgumentException("Cannot multiply $this and $visit")
    }
}

private operator fun Any?.div(visit: Any?): Any? {
    return if (this is Byte && visit is Byte) {
        this / visit
    } else if (this is Short && visit is Short) {
        this / visit
    } else if (this is Int && visit is Int) {
        this / visit
    } else if (this is Long && visit is Long) {
        this / visit
    } else if (this is BigInteger && visit is BigInteger) {
        this.divide(visit)
    } else if (this is BigDecimal && visit is BigDecimal) {
        this.divide(visit)
    } else {
        throw IllegalArgumentException("Cannot add $this and $visit")
    }
}
